### PR TITLE
SignatureDrop: using `DropSinglePhase`

### PR DESCRIPTION
**Pull request by nkrishang**

### CODE REVIEW

File: contracts/feature/interface/IDropSinglePhase.sol [hunk 1]
```diff
@@ -10,11 +10,11 @@ interface IDropSinglePhase is IClaimCondition {
     }
 
     event TokensClaimed(
-        ClaimCondition condition,
+        uint256 indexed claimConditionIndex,
```
Comment (jakeloo) [inline contracts/feature/interface/IDropSinglePhase.sol:13]: probably doesn't need to be indexed. 
single phase = index always 0?

File: contracts/signature-drop/SignatureDrop.sol [hunk 2]
```diff
@@ -202,50 +188,31 @@ contract SignatureDrop is
...
 
         uint256 tokenIdToMint = _currentIndex;
         require(tokenIdToMint + _req.quantity <= nextTokenIdToMint, "not enough minted tokens.");
 
+        // Verify and process payload.
+        signer = _processRequest(_req, _signature);
+
         // Get receiver of tokens.
-        address receiver = _req.to == address(0) ? msg.sender : _req.to;
+        address receiver = _req.to == address(0) ? _msgSender() : _req.to;
```
Comment (jakeloo) [inline contracts/signature-drop/SignatureDrop.sol:205]: we should probably inform people that this is vulnerable to signature frontrunning attack

**Comment by jakeloo** [comment]
https://github.com/chiru-labs/ERC721A/blob/main/contracts/ERC721A.sol#L742-L744 is not overriden

**Comment by jakeloo** [comment]
https://github.com/thirdweb-dev/contracts/blob/7f575f5e30b97cc318941e88a756208255db5b8e/contracts/signature-drop/SignatureDrop.sol#L96-L110
```
        _setupRole(DEFAULT_ADMIN_ROLE, _msgSender());

        setContractURI(_contractURI);
        setOwner(_defaultAdmin);

        _setupRole(DEFAULT_ADMIN_ROLE, _defaultAdmin);
        _setupRole(MINTER_ROLE, _defaultAdmin);
        _setupRole(TRANSFER_ROLE, _defaultAdmin);
        _setupRole(TRANSFER_ROLE, address(0));

        setPlatformFeeInfo(_platformFeeRecipient, _platformFeeBps);
        setDefaultRoyaltyInfo(_royaltyRecipient, _royaltyBps);
        setPrimarySaleRecipient(_saleRecipient);

        _revokeRole(DEFAULT_ADMIN_ROLE, _msgSender());
```

`_setupRole` docstring:
```
This function should only be called from the constructor when setting up the initial roles for the system. Using this function in any other way is effectively circumventing the admin system imposed by AccessControl.
```

i think this setup is a little weird and wasteful. basically with `_setupRole`, you don't need to be an admin to do so.

Other thought, could we have `_setupX` internal functions for our setters to address the same problem?

File: contracts/signature-drop/SignatureDrop.sol [hunk 3]
```diff
@@ -325,21 +297,18 @@ contract SignatureDrop is
...
-        address ownerOfToken = ownerOf(tokenId);
-        //solhint-disable-next-line max-line-length
-        require(
-            _msgSender() == ownerOfToken ||
-                isApprovedForAll(ownerOfToken, _msgSender()) ||
-                getApproved(tokenId) == _msgSender(),
-            "caller not owner nor approved"
-        );
-        _burn(tokenId);
+        _burn(tokenId, true);
```
Comment (jakeloo) [inline contracts/signature-drop/SignatureDrop.sol:311]: make a comment here that `_burn` internally check for approvals from erc721a

Comment (nkrishang) [reply contracts/signature-drop/SignatureDrop.sol:311]: Will add consideration in a code comment + design document.

**Comment by nkrishang** [comment]
> https://github.com/chiru-labs/ERC721A/blob/main/contracts/ERC721A.sol#L742-L744 is not overriden

We were using `erc721a-upgradeable` v3.3 where this `_msgSenderERC721A` did not exist. Will upgrade to the latest `erc721a-upgradeable` v4.0 where `_msgSenderERC721A` does exist, and will override it.

**Pull request merged**